Review: OAK-8827 AWS support for segment-tar

synox opened the pull request:
Do not merge.

Commit:
OAK-8827 - AWS support for segment-tar
Journal file implementation for AWS

Commit:
OAK-8827 - AWS support for segment-tar
repository lock implementation using dynamodb-lock-client

Commit:
OAK-8827 - AWS support for segment-tar
Project folder oak-segment-aws and POM

Commit:
OAK-8827 - AWS support for segment-tar
Implement ManifestFile

Commit:
OAK-8827 - AWS support for segment-tar
GCJournal file backed by DynamoDB

Commit:
OAK-8827 - AWS support for segment-tar
SegmentWriteQueue for sending archive entries to AWS

Commit:
OAK-8827 - AWS support for segment-tar
Archive manager and Persistence

Commit:
OAK-8827 - AWS support for segment-tar
Fixture for AWS and integration tests

Commit:
OAK-8827 - AWS support for segment-tar
SegmentCopyCommand capability to copy between AWS and TAR persistence

Commit:
OAK-8827 - AWS support for segment-tar
CompactCommand support for AWS

Commit:
OAK-8827 - AWS support for segment-tar
AwsSegmentStoreService to setup persistence for custom store

synox commented on oak-segment-aws/src/test/java/org/apache/jackrabbit/oak/segment/aws/tool/SegmentCopyAzureToTarTest.java:
```diff
@@ -0,0 +1,67 @@
+import org.junit.Before;
+import org.junit.ClassRule;
+
+public class SegmentCopyAzureToTarTest extends SegmentCopyTestBase {
```
Should be SegmentCopy**Aws**ToTarTest

synox commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/Configuration.java:
```diff
@@ -0,0 +1,73 @@
+    @AttributeDefinition(
+            name = "Root directory",
+            description = "Names of all the created blobs will be prefixed with this path")
+    String journalTableName() default AwsSegmentStoreService.DEFAULT_JOURNALTABLE_NAME;
```
Missing definition. 
```java
            name = "AWS DynamoDB journal table name",
            description = "Name of table used for storing log entries for journal and gc.")
```

synox commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/Configuration.java:
```diff
@@ -0,0 +1,73 @@
+    @AttributeDefinition(
+            name = "Root directory",
+            description = "Names of all the created blobs will be prefixed with this path")
+    String lockTableName() default AwsSegmentStoreService.DEFAULT_LOCKTABLE_NAME;
```
Missing definition. 
```java
            name = "AWS DynamoDB lock table name",
            description = "Name of table used for managing the distributed lock.")
```

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsAppendableFile.java:
```diff
@@ -0,0 +1,132 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AwsAppendableFile {
```
One question related to this class: why not include its functionality as part of `AwsJournalFile` and get rid of it altogether? To me it doesn't make sense to keep it around.

alvarorahul replied:
agreed, removed this file.

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsArchiveManager.java:
```diff
@@ -0,0 +1,207 @@
+
+    @Override
+    public SegmentArchiveReader open(String archiveName) throws IOException {
+        AwsContext directoryContext = awsContext.withDirectory(archiveName);
```
`AwsContext` has two responsibilities: providing the overall context for all AWS operations and also more granular context relative to an archive/directory in S3.. I think it's a bit misleading and makes also some random parts of the code harder to read.

alvarorahul replied:
This has been split into a `DynamoDBClient` class for handling operations on tables and `S3Directory` class for handling operations within a specific directory in an S3 bucket

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsContext.java:
```diff
@@ -0,0 +1,441 @@
+            rootDirectory = rootDirectory.substring(1);
+        }
+
+        AWSCredentials credentials = configuration.sessionToken() == null || configuration.sessionToken().isEmpty()
```
Per Miroslav's comments: 

General security reccomendation in AWS, when consuming services (S3 and DynamoDB in this case) is to prefer using temporary security credentials instead of long-term access keys. 
https://docs.aws.amazon.com/general/latest/gr/aws-access-keys-best-practices.html#use-roles

Seems that class `AwsContext.java` is using long-term access keys.
Link below explains how AWS IAM roles can be associated with k8n service account. 
https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html

Credentials provider that is used currently is fine during development/testing.

alvarorahul replied:
It supports both. Long-term access is obtained by using accessKey and secretKey of either the root user or an IAM user.

Temporary security credentials work by using a sessionToken in conjunction with accesKey and secretKey. More details here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials

The current implementation does support sessionToken and uses it when specified.

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsContext.java:
```diff
@@ -0,0 +1,441 @@
+
+        AmazonS3 s3 = AmazonS3ClientBuilder.standard().withCredentials(credentialsProvider).withRegion(region)
+                .withRequestHandlers(handler).build();
+        AmazonDynamoDB ddb = AmazonDynamoDBClientBuilder.standard().withCredentials(credentialsProvider)
```
Is it ok to re-use the previous hander here? Our request metrics will be skewed b/c frequent calls to update journal...

alvarorahul replied:
this matches what we're doing in oak-segment-azure - every journal update is a blob append operation which results in a network call and that gets logged

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsContext.java:
```diff
@@ -0,0 +1,441 @@
+                .with(TABLE_ATTR_CONTENT, line);
+        try {
+            try {
+                Thread.sleep(1L);
```
I don't get this part: what's the purpose of the call to `Thread.sleep`?

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsGCJournalFile.java:
```diff
@@ -0,0 +1,46 @@
+
+import org.apache.jackrabbit.oak.segment.spi.persistence.GCJournalFile;
+
+public class AwsGCJournalFile implements GCJournalFile {
```
I think this is not correctly implemented as it uses an `AwsJournalFile` under the hood
// the problem is that the entries need not be in reverse order as imposed by `JournalFileReader` interface.

alvarorahul replied:
it does do the right thing... though I agree... using JournalFileReader makes this confusing... I've removed AwsAppendableFile.java altogether

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsJournalFile.java:
```diff
@@ -0,0 +1,52 @@
+
+public class AwsJournalFile implements JournalFile {
+
+    private final AwsAppendableFile file;
```
are the entries returned in LIFO order? also, can't this class directly implement the functionality provided by `AwsAppendableFile`? See my previous comment on that class.

alvarorahul replied:
removed AwsAppendableFile.java

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsPersistence.java:
```diff
@@ -0,0 +1,98 @@
+
+    protected final AwsContext awsContext;
+
+    private final String fileNameSuffix;
```
I couldn't understand what's this suffix and how it's used...

alvarorahul replied:
This is necessary to handle the IT where it spins up multiple stores are created in the setUpCluster method and we use a single bucket for that.. see changes to
oak-run-commons/src/main/java/org/apache/jackrabbit/oak/fixture/SegmentTarFixture.java

UPDATE: changed the IT to use multiple buckets and removed this suffix

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsPersistence.java:
```diff
@@ -0,0 +1,98 @@
+        this(awsContext, null);
+    }
+
+    public AwsPersistence(AwsContext awsContext, String id) {
```
The same applies for this id.

alvarorahul replied:
same comment as above... in the setUpCluster method multiple stores are setup this suffix is used to identify the entries for journal files for each store in the cluster

UPDATE: the IT has been updated and id and fileNameSuffix have been removed altogether

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsRepositoryLock.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static final Logger log = LoggerFactory.getLogger(AwsContext.class);
+
+    private static final int TIMEOUT_SEC = Integer.getInteger("oak.segment.azure.lock.timeout", 0);
```
This should be `oak.segment.aws.lock.timeout`.

alvarorahul replied:
updated

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsSegmentStoreService.java:
```diff
@@ -0,0 +1,70 @@
+@Component(configurationPolicy = ConfigurationPolicy.REQUIRE, configurationPid = { Configuration.PID })
+public class AwsSegmentStoreService {
+
+    public static final String DEFAULT_BUCKET_NAME = "oak";
```
unused constant

alvarorahul replied:
removed

dulceanu commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/Configuration.java:
```diff
@@ -0,0 +1,73 @@
+
+@ObjectClassDefinition(
+        pid = {PID},
+        name = "Apache Jackrabbit Oak Azure Segment Store Service",
```
Azure needs to be replaced with AWS.

alvarorahul replied:
updated

dulceanu reviewed (changes requested):
Finished 1st pass through the code.

rootpea committed:
Change AWS metadata keys to lowercase

https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html
Amazon S3 stores user-defined metadata keys in lowercase.

synox commented on oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsArchiveManager.java:
```diff
@@ -0,0 +1,207 @@
+
+    @Override
+    public void recoverEntries(String archiveName, LinkedHashMap<UUID, byte[]> entries) throws IOException {
+        Pattern pattern = Pattern.compile(SEGMENT_FILE_NAME_PATTERN);
```
The Pattern object should be a final static constant of the class.

synox reviewed (commented):
Please also update the docs for oak-run: https://github.com/alvarorahul/jackrabbit-oak/blame/alvarorahul/oak-segment-aws/oak-doc/src/site/markdown/nodestore/segment/overview.md#L673

Commit:
not all bytes being read from segment resulting in SegmentNotFoundException

Commit:
Address review comments and update versions of dependencies

Commit:
Split AwsContext into DynamoDBClient and S3Directory classes

Commit:
Update version of oak-segment-aws to current iteration

Commit:
Fix Publish AWS warnings

Commit:
Use timeout value specified in config for the distributed lock

Commit:
OAK-8827 - Handle revisions and consume 'last' parameter in segmentcopycommand

Commit:
OAK-8827 - AWS support for segment-tar
Track requests, errors and duration for AWS calls

Commit:
OAK-8827 - fix for listObjects not following continuation tokens

Commit:
AWS oak-run: class not found: org/apache/commons/lang3/StringUtils

Commit:
Attempt at fixing performance issue when opening AwsSegmentArchiveReader

Commit:
Update markdown in oak-doc for oak-segment-aws
